### PR TITLE
HDDS-7829. SCM to reject adding container to closed pipeline

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

This is a correction of the fix for "SCM terminates when adding containers to a closed pipeline" in [HDDS-7738](https://issues.apache.org/jira/browse/HDDS-7738).

In general, due to the lack of atomicity and the fact that SCM status can change between transactions, SCM should just reject transactions gracefully if some logical condition is not met.

For example, when processing a transaction adding a container, SCM expects the associated pipeline to be in an `OPEN` state and rejects the container creation if the pipeline is `CLOSED`.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7829

## How was this patch tested?
Unit test, integ test.

### CODE REVIEW

**Comment by kerneltime** [comment]
@swamirishi can you please take a look?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 1]
```diff
@@ -136,7 +137,14 @@ public CompletableFuture<Message> applyTransaction(
     try {
       final SCMRatisRequest request = SCMRatisRequest.decode(
           Message.valueOf(trx.getStateMachineLogEntry().getLogData()));
-      applyTransactionFuture.complete(process(request));
+
+      try {
+        applyTransactionFuture.complete(process(request));
+      } catch (SCMException ex) {
+        // SCMException is considered as logical rejection and is returned to
+        // Ratis client.
```
Comment (errose28) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:145]: From `SCMException#ResultCodes`, `IO_EXCEPTION` and `INTERNAL_ERROR` should probably still crash the SCM

**Review by errose28 — COMMENTED** [review]
Thanks for fixing this @duongkame. If you have time it might be good to look around the code and see if there are any other instances where IOException is being thrown instead of SCMException to proactively catch future problems like this.

I feel like there should be a way from integration tests to make the SCM leader send close pipeline and container create commands for that pipeline back to back to test this case. I can help investigate this.

**Comment by errose28** [comment]
Following up on the testing part of this patch, the unit test is good, but it does not test the changes to `SCMStateMachine#applyTransaction`. It looks like this could be done by modifying `TestPipelineClose#testPipelineCloseWithClosedContainer` to use 1 node Ratis for SCM, then calling `pipelineManager#addContainerToPipeline` on the closed pipeline and check that an exception is thrown.

I don't see a way to test that generic IOExceptions or SCMExceptions with fatal result codes terminate the SCM without failing the test due to JVM exit.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java (hunk 1 above)
Comment (duongkame) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java]: Good point, updated.

**Comment by duongkame** [comment]
> Following up on the testing part of this patch, the unit test is good, but it does not test the changes to `SCMStateMachine#applyTransaction`. It looks like this could be done by modifying `TestPipelineClose#testPipelineCloseWithClosedContainer` to use 1 node Ratis for SCM, then calling `pipelineManager#addContainerToPipeline` on the closed pipeline and check that an exception is thrown.

@errose28, thanks to your suggestion, I was able to create a proper integration test for the failure scenario.

> I don't see a way to test that generic IOExceptions or SCMExceptions with fatal result codes terminate the SCM without failing the test due to JVM exit.

I believe that can be resorted by `ExitUtils.disableSystemExit`, which prevents actual JVM exit for testing purposes.
Yet, I scanned the code and could not find a good way to produce internal errors, maybe we can somehow generate `TimeoutException` and SCM always translates it to internal errors.

**Review by errose28 — APPROVED** [review]
Thanks @duongkame LGTM

**Pull request merged**